### ISSUE TITLE
Phone number sign-in doesn't work in Cordova

### ISSUE THREAD

**Issue by FluorescentHallucinogen** [firebase/firebase-js-sdk#102]
Phone authentication is not supported in Cordova environment due to the limitation in the underlying Firebase core SDK.

**Comment by bojeil-google** [comment]
Yeah, this is a limitation now as we need to apply some origin checking before sending an SMS message. Origins in Cordova apps look like: file://assets/.../index.html. We won't be able to tell a legitimate request from a fraudulent while guarding against possible abuse or phishing attacks.

A workaround is provided here: https://stackoverflow.com/questions/44081040/ionic2-authentication-firebase

We are looking into providing ready support for something of the like but we can't guarantee anything at the moment or any timeline for this.

**Comment by washowasho** [comment]
@bojeil-google, the work-around using a browser tab + universal links works well with Android... but it does not with iOS. I am working on figuring out why. The browser tab does not auto-close on the iOS side after recaptcha is complete.

**Comment by bojeil-google** [comment]
browsertab plugin has a close method for iOS: cordova.plugins.browsertab.close
It is not well documented. Once you intercept the link or the app is visible again, close the browser tab.

**Comment by washowasho** [comment]
@bojeil-google , yes I am aware of the close() method... however I do not see how the browser tab plugin can "intercept" the URL in order to invoice the close() method. Am I missing something?

**Comment by bojeil-google** [comment]
you have to use cordova-universal-links-plugin to intercept a universal link.

**Comment by washowasho** [comment]
@bojeil-google That's actually where I am having the trouble now. I was not successful in getting the cordova-universal-links-plugin working with iOS.

**Comment by bojeil-google** [comment]
I speculate this is a problem in your universal link implementation. We use that for Cordova OAuth flow in Firebase Auth with Firebase Dynamic links. Try using FDL for your iOS universal links.

**Comment by washowasho** [comment]
Yes, I am using FDL as well. I notice the browsertab works well with the OAuth flow, and I was hoping to recreate it the same way. Do you have a sample project or a link to the code that shows this?

**Comment by bojeil-google** [comment]
Oh actually we don't use FDL/universal links for iOS. We use custom scheme redirects there. But it should still work for iOS. I tested that back in the day. I don't have sample code as we never actually used it for iOS. Make sure you are constructing the FDL link correctly in your app. Also make sure you are registering you iOS app to detect this universal link. Add the <universal-links> configuration in config.xml. Then add the subscriber to detect the link:
`universalLinks.subscribe('eventName', function (eventData) {})`

Try testing without the Firebase js library to make sure it is not swallowing the income link or something.

**Comment by washowasho** [comment]
@bojeil-google Thank you. I did all of what you mentioned, unfortunately the universal link (UL) does not work. The weird thing is... if I execute the UL from Chrome on iOS... it opens the app just fine. Safari... not so much. Even though the custom URL scheme doesn't seem to be supported much by Apple these days, I may be taking that route as well.

**Comment by washowasho** [comment]
@bojeil-google I was able to get Universal Links working on both iOS and Android, but Android opens up the WebView/BrowserTab outside of the app, which basically makes it impossible to close (iOS is cleaner because the BrowserTab opens within the app). However, I do see how "Cordova OAuth flow in Firebase Auth with Firebase Dynamic" manages to close the extra window? Can you enlighten us on how closing the extra BrowserTab is accomplished?

**Comment by bojeil-google** [comment]
In Android, an opened chrome custom tab via browser tab plugin will close itself when an app link is clicked within it. That is the behavior of the Cordova OAuth flow. In the case where Chrome custom tab is unavailable, a regular browser window is used. You can't close a browser window from the app but it is not necessary to close it in that case. The link click should redirect to the app. You can add some delay in the browser window to redirect to a blank page. A similar behavior is used in the Cordova OAuth flow.

**Comment by washowasho** [comment]
Thank you for the response. Looks like my page is loading in a new browser instead of the browser tab. 

Finally good with iOS! Had to settle with custom URL scheme for now.

**Comment by washowasho** [comment]
Ok, the best way to control browsertab closing in Android is using `<preference name="AndroidLaunchMode" value="singleTask" />`. @bojeil-google, this closes automatically when in this mode. I had it set to singleInstance.

So all-in-all, using a Browser Tab to re-direct back to app after typing in the verification code in HTTPS source site seems to be an acceptable work around! Thanks for your help @bojeil-google .

Since it seems RECAPTCHA only works when on HTTPS, how is Firebase going to ultimately allow for this inside of the Cordova app as opposed to relying on Browser tab?

**Comment by bojeil-google** [comment]
I am not sure what you mean. Browser tab will open a chrome custom tab or SFSafariViewController where you then open a web page there that is whitelisted in your authorized domains (doesn't have to be HTTPS though you get that for free with Firebase hosting and I highly recommend using that) and render a reCAPTCHA there.
Any solution in Cordova will have to rely on a similar mechanism. This is unavoidable unless you wrap the underlying native SDKs which rely on a different mechanism for anti-abuse/verification.

**Comment by derwaldgeist** [comment]
I have implemented phone number auth using MeteorJS, where the app is actually served from a local server (http://meteor.local) running on the device.

In principle, this works. But sometimes I get weird "network error" messages, especially on Android. I allowed access to google.com, googleapis.com and gstatic.com, for both http and https, but still the error appears. No errors on the console, though.

Am I missing something?
